nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/StreamBulkScroller.java

```diff
@@ -135,50 +137,45 @@ protected void processRecord(ComputationContext context, Record record) {
                     long documentCount = 0;
                     while (scroll.hasResults()) {
                         List<String> docIds = scroll.getResults();
-                        // send these ids as keys to the appropriate stream
-                        // key will be bulkOperationId/docId
-                        // value/data is a BulkCommand serialized as JSON
-                        docIds.forEach(docId -> context.produceRecord(command.getOperation(), record.key + '/' + docId,
-                                record.data));
+                        documentIds.addAll(docIds);
+                        if (documentIds.size() >= recordBatchSize) {
+                            // send these ids as keys to the appropriate stream
+                            // key will be bulkActionId/docIds
+                            // value/data is a BulkCommand serialized as JSON
```

Comment:
oh the key is the list of ids and the value the bulk command.
the key is limited is not supposed to contain complex or long data, the main usage of the key is to shard records, in your case a simple roundrobin sharding is enough the key can be "%s:%s", actionId, count.
the value which is the record shoudl contains the action id and the list of doc ids.
let's disccuss about this.

Reply:
count was supposed to be incremented for each batch, the result will be the same as using a randomUUID without overhead